Improve the TextMate grammar

qligier opened the pull request:
This improves the TextMate grammar. It fixes the coloring of stars in cardinality and add support for:

- cardinality, alias and caret path;
- _reference_/_canonical_ content;
- declaration name;
- _from_, _obeys_ and _insert_ rules.

Before:
![before](https://user-images.githubusercontent.com/3299399/198830258-f89992ba-90ee-4824-a421-87094880e7d6.png)

After:
![after](https://user-images.githubusercontent.com/3299399/198830267-700033ee-689d-4b9c-8699-d7e229b426c5.png)

qligier commented:
Converting to draft, I think I messed up cardinalities like `0..`.

qligier commented:
Incomplete cardinalities are properly supported now.

jafeltra commented:
Hi @qligier. Apologies for the long delay on this PR! We really appreciate the contribution, and hopefully we can get this in soon. It is really helpful to have so many additional entities highlighted. In testing the highlighting, I noticed a few places that should be updated before merging this PR:

- There's a bug in the highlighting of declaration names with `-` in them. Something like `Instance: mcode-receiver-patient-bundle` only highlights the first part (mcode) of the declaration name. It would be good to fix this before merging in these updates.
- There's a bug around rules with cardinalities that have a comment after them. If there is a rule such as `* subject 1..1 // This is a comment`, the comment is only highlighted correctly if there are two spaces between the end of the cardinality and the comment. This should also be fixed before merging.
- There is another bug with highlighting Insert Rules that insert parameterized rule sets with open and escaped close parentheses. A rule like `* insert MapConcept(#15825003, "Aortic structure (body structure\)")` does not account for the `\)` and the syntax highlighting is off beyond it. This particular bug is a little tricky because we will be introducing new syntax in the next major version of SUSHI that will support a syntax that does not require escaping the `)` used in a parameter. With this syntax, the insert rule can look like `* insert MapConcept(#15825003, [["Aortic structure (body structure)"]])`. If this PR can support the new `[[ ]]` syntax as well, that would be a bonus. However, if just the `\)` can be handled so that the syntax highlighting works as expected that is sufficient for this PR. Alternatively, this PR can revert all syntax handling inside the Rule Set parameters completely so they are highlighted as they are in the published version of the extension. Any of these options will work.
- I think it would be good to include the `Id` keyword in the same group as `Profile` and `Parent` so that the Id's value is also highlighted.
- It would be nice to have the Caret Value path highlighted in a color that is different from the one we are using for numbers. I experimented with a few different colors and it seemed like setting the `name` as `"constant.character.escape.fsh"` uses a distinct color.
- It would be nice to have the Type rules with `only` highlight the datatype that follow it, similar to how `insert`/`from`/`obeys` highlight the entity that comes after them. For example, a rule like `* effective[x] only dateTime` would highlight `dateTime`. Ideally, this wouldn't change how a Reference that follows an `only` is highlighted because I like how rules like `* performer only Reference(USCorePractitioner)` use the common `Reference()` highlight.

I recognize this PR was contributed many months ago, and you may not have the time to dedicate to this PR anymore. If you'd prefer, I'm happy to make these changes.

qligier commented:
Hi @jafeltra! Thank you for testing the modifications.

1. Fixed with efdb405c3ab2b8fd0d9dba502850ea161990e28f.
2. Fixed with b58c6d766747ee4967f7206cb24d2b21ba8243d5, plus b3f23cd4c3817c1e65a35d4f339dac87df6dc846 that fixes the missing highlighting when a single-line comment directly follows something else, without spaces between.
3. Fixed with 4f9bb4148116b6b2a2ed48d42ed4ba008ae07744. Supporting the new syntax requires more work, so it's not done yet.
4. Fixed with 9094152a0782688c824d52017704be1fe662c987.
5. Fixed with 795ab23b8f1d0a5facb7d615f983bfb1f9e81bac.
6. Fixed with 78459ce450b59d362a451fbb95d1fba3b4211437.

Preview:
![image](https://user-images.githubusercontent.com/3299399/236669263-2bd4b11c-f619-4aaf-a288-12bfd3770c61.png)

jafeltra commented on syntaxes/fsh.tmLanguage.json:
```diff
@@ -29,11 +29,91 @@
+          "match": "\\b(Alias|Expression|Description|Mixins|Severity|Target|Title|Usage|XPath)(?=\\s*:)\\b"
+        },
+        {
+          "match": "\\b(CodeSystem|Extension|Id|Instance|InstanceOf|Invariant|Logical|Mapping|Parent|Profile|Resource|RuleSet|Source|ValueSet)\\s?(:)\\s+([A-Za-z0-9_-]+)\\b",
```
This set of characters reminded me that you can also include `.` in some of the values after these keywords, so I think it would be good to include the `.` character with in the `([A-Za-z0-9_-]+)` set here.

jafeltra reviewed (commented):
Thanks @qligier for the quick response and for addressing all of my comments! I tested everything and it is looking really great. The list of each commit was super helpful as well. And thanks for fixing [`b3f23cd`](https://github.com/standardhealth/vscode-language-fsh/commit/b3f23cd4c3817c1e65a35d4f339dac87df6dc846)! I'm not sure how I never noticed that not working before.

I only noticed one other very small thing. I left an inline comment, but essentially the `.` character is also allowed in some ids that follow some of the FSH keywords, so I suggested including that in the set of characters to follow.

jafeltra commented:
> > And thanks for fixing [b3f23cd](https://github.com/standardhealth/vscode-language-fsh/commit/b3f23cd4c3817c1e65a35d4f339dac87df6dc846)! I'm not sure how I never noticed that not working before.
> 
> Side note, there are some corner cases where a white space is required before a line comment, as in `InstanceOf: http://hl7.org//This is a comment`. I am not sure if something could be done in the FSH grammar to improve that.

Hmm that's a good point. I guess that might it a little confusing that VS Code will now highlight that portion as if it were a comment. I can discuss with the team, and if it is a restriction that is best kept in the FSH grammar, we can consider reverting that particular commit.

jafeltra commented:
> > And thanks for fixing [b3f23cd](https://github.com/standardhealth/vscode-language-fsh/commit/b3f23cd4c3817c1e65a35d4f339dac87df6dc846)! I'm not sure how I never noticed that not working before.
> 
> Side note, there are some corner cases where a white space is required before a line comment, as in `InstanceOf: http://hl7.org//This is a comment`. I am not sure if something could be done in the FSH grammar to improve that.

After thinking about it a little more and chatting with another member of the FSH team, I think we actually just want to revert [`b3f23cd`](https://github.com/standardhealth/vscode-language-fsh/commit/b3f23cd4c3817c1e65a35d4f339dac87df6dc846). The limitation in the FSH grammar is something that we'll need to maintain since it allows URLs to be properly parsed. So I think keeping the start of the regex for comment highlighting the same here makes the most sense.

Sorry about the back and forth!

qligier commented:
I don't mind reverting b3f23cd, but I don't understand why. What is the FSH grammar limitation we are talking about?
To be clear, something like `InstanceOf: http://hl7.org//Comment` is parsed as a single URL even with b3f23cd, the commit helps in other cases like `* subject 1..1//Comment`.

jafeltra commented:
> something like `InstanceOf: http://hl7.org//Comment` is parsed as a single URL even with [b3f23cd](https://github.com/standardhealth/vscode-language-fsh/commit/b3f23cd4c3817c1e65a35d4f339dac87df6dc846), the commit helps in other cases like `* subject 1..1//Comment`.

The problem is that, even though something like `* subject 1..1//Comment` is highlighted as if `//Comment` is a comment in VS Code with [`b3f23cd`](https://github.com/standardhealth/vscode-language-fsh/commit/b3f23cd4c3817c1e65a35d4f339dac87df6dc846), the FSH grammar and SUSHI will not parse `//Comment` as a comment and will actually cause an error when running SUSHI on that particular FSH line. Because the FSH grammar is expecting the space before the `//`, it does not treat it as a comment, which causes SUSHI parsing errors. This limitation to have the whitespace before the comment is in the grammar because there are other cases where it would be unclear if the `//` is part of the value used in the FSH rule or are the start of a comment.

While this is something we could consider revisiting in the grammar itself, I don't think we want VS Code to highlight this case as if it were a comment while SUSHI does not parse it that way. I think it will be confusing to users when they get an error but it looks as though their comment is being processed as a comment in VS Code.

qligier commented:
Okay, I understand. I only read the line `LINE_COMMENT: '//' ~[\r\n]* [\r\n] -> skip;` in the lexer and thought a whitespace was not necessary, I did not try sushi on the flawed example.
Then I will remove that commit and clean the history.

qligier commented:
I have removed b3f23cd from the history.

jafeltra commented:
@qligier - I just published a new version of the FSH extension, [version 1.11.0](https://marketplace.visualstudio.com/items?itemName=MITRE-Health.vscode-language-fsh), which includes these changes. It should be available in the Marketplace soon, if not already. Thanks again for the contribution and for the timely updates this past week!
